### DISCUSSION TITLE
Is there a helper to build a tree from a vec of file paths in `gix` ?

### DISCUSSION THREAD

**Discussion by dlight** [GitoxideLabs/gitoxide#1160] [Q&A] [answered]
Right now I'm starting with an empty tree, writing file contents as blobs and adding them to the tree one by one. This has proven to be somewhat complicated as it needs to build directories recursively, needs to handle symlinks correctly, etc.

Is there a higher level `gix` function to do this?

**Comment by Byron** [comment, marked as answer]
No, there is no such functionality available yet.

You can probably subscribe to the respective tracking ticket or check back from time to time: https://github.com/Byron/gitoxide/issues/301 .

Please also note that this is something I will eventually work on [as it's also required by `cargo`](https://github.com/Byron/gitoxide/issues/106).

**Reply by dlight** [reply]
Good to know, thanks!

Just to be clear, my question is about building a tree with arbitrary paths (for example, without necessarily including a whole subdirectory, but perhaps just a few files from it), without touching the index at all.

Not touching the index is perhaps an unusual demand (not sure), but it is necessary to not disrupt the workflow of someone that is interacting with the same repository concurrently through the git cli.

**Reply by Byron** [reply]
The index, in `git`, is the default way to create new trees and to do so quickly.

Not changing it would be as easy as not writing it back to disk, but discard the in-memory changes after usage.

Right now, while everything is done by hand anyway, nothing stops you from working directly with the object database though.

**Reply by dlight** [reply]
My issue is that perhaps the index already have some stuff manually added with `git add`; I don't want to touch that (because I'm writing a tree unrelated to whatever is happening right now in the repository). You mean I can write to an in-memory index, unrelated to the `.git/index` file? That would be ok for me.

But `write_object` from [the `Repository`](https://docs.rs/gix/latest/gix/struct.Repository.html#method.write_object) or even [from `CommitAutoRollback`](https://docs.rs/gix/latest/gix/config/struct.CommitAutoRollback.html#method.write_object) will not use `.git/index`, but an in-memory index? Can I control that? (I suppose I could make a new question for that)

**Reply by Byron** [reply]
>  You mean I can write to an in-memory index, unrelated to the .git/index file? That would be ok for me.

You can [create an index from a tree](https://docs.rs/gix/latest/gix/struct.Repository.html#method.index_from_tree) with ease, but as said before, without the ability to create a tree from an index using an index has no benefit.

> But write_object from [the Repository](https://docs.rs/gix/latest/gix/struct.Repository.html#method.write_object) or even [from CommitAutoRollback](https://docs.rs/gix/latest/gix/config/struct.CommitAutoRollback.html#method.write_object) will not use .git/index, but an in-memory index?

Nothing uses an index implicitly.

Maybe `git2` is a better choice for you at this time, as it has all the facilities you'd require.
